### PR TITLE
Factor code between records and typeclasses

### PR DESCRIPTION
**Pull request by proux01**
This is a followup of https://github.com/coq/coq/pull/12611 factoring some code between record and typeclass declarations.

#### Overlay (to be merged in sync with the PR):
* https://github.com/LPCIC/coq-elpi/pull/357

### CODE REVIEW

File: vernac/record.ml [hunk 1]
```diff
@@ -592,20 +593,34 @@ let declare_structure ~cumulative finite ~univs ~variances ~primitive_proj
...
 
-let build_class_constant ~univs ~rdata ~primitive_proj field implfs params paramimpls coers binder id proj_name =
+(* declare definitional class (typeclasses that are not record) *)
+let declare_class_constant ~univs paramimpls params data =
+  let id, rdata, coers, inhabitant_id = match data with
+    | [{ Data.id; rdata; is_coercion; coers; inhabitant_id }] ->
+       assert (not is_coercion);  (* should be ensured by caller *)
+       let coers = List.map (fun { pf_subclass; pf_priority } -> if pf_subclass then Some { hint_priority = pf_priority; hint_pattern = None } else None) coers in
+       id, rdata, coers, inhabitant_id
+    | _ -> assert false in  (* should be ensured by caller *)
```
Comment (ejgallego) [inline vernac/record.ml:616]: Not a fan of this, why are we taking a list in the first place?

**Comment by ejgallego** [comment]
Sorry @proux01 , I'm having trouble following why this refactoring is an improvement. Could you summarize the idea a bit?

**Comment by proux01** [comment]
@ejgallego the goal is to make more apparent what's common (and what differs) between record and class declarations. In particular the call to `declare_structure` that is common to both codes but was deeply burried in the class declaration functions.

**Comment by proux01** [comment]
There was even a comment in the previous code:
```OCaml
(* declaring structures, common data to refactor *)
```

File: vernac/record.ml (hunk 1 above)
Comment (proux01) [reply vernac/record.ml:616]: Agree, this code is duplicated and should go in a specific function. The list comes from the fact that record (unlike typeclasses) can be mutually recursive records (with the `Inductive ... := { ... } with ... := { ... }` syntax.

**Comment by ejgallego** [comment]
Thanks, I'll have a look again hopefully soon.

**Comment by ejgallego** [comment]
Hi @proux01 sorry for the delay, can you rebase?

File: vernac/record.ml [hunk 2]
```diff
@@ -413,7 +414,7 @@ let build_proj env mib indsp primitive x rp lifted_fields ~poly paramdecls param
 
 (** [declare_projections] prepares the common context for all record
    projections and then calls [build_proj] for each one. *)
-let declare_projections indsp univs ?(kind=Decls.StructureComponent) inhabitant_id flags fieldimpls fields =
+let declare_projections indsp univs ~kind inhabitant_id flags fieldimpls fields =
```
Comment (Alizter) [inline vernac/record.ml:417]: Maybe rename kind to proj_kind here too.

File: vernac/record.ml [hunk 3]
```diff
@@ -692,29 +692,34 @@ let build_record_constant ~rdata ~univs ~variances ~cumulative ~template ~primit
...
+    | [{ Data.id; idbuild; rdata; is_coercion; coers; inhabitant_id }] ->
+       if is_coercion then warn_future_coercion_class_constructor ();
+       if List.exists (fun pf -> pf.pf_subclass) coers then
+         warn_future_coercion_class_field ();
+       let coers = List.map (fun { pf_subclass; pf_priority } -> if pf_subclass then Some { hint_priority = pf_priority; hint_pattern = None } else None) coers in
+       id, idbuild, rdata, is_coercion, coers, inhabitant_id
     | _ ->
-      build_record_constant ~rdata ~univs ~variances ~cumulative ~template ~primitive_proj
-        fields params paramimpls is_coercion coers id idbuild inhabitant_id
+       CErrors.user_err (str "Mutual definitional classes are not handled.") in
```
Comment (Alizter) [inline vernac/record.ml:708]:
```suggestion
       CErrors.user_err (str "Mutual definitional classes are not supported.") in
```

Comment (Alizter) [reply vernac/record.ml:708]: Just a suggestion, I know it was worded like this before.

File: vernac/record.ml (hunk 2 above)
Comment (proux01) [reply vernac/record.ml]: Not sure about that one, the function is already called `declare_projection`, so I'd find `proj_kind` a bit redundant (and that wouldn't simplify calls since the argument when calling it is called `projection*s*_kind`).

**Review by Alizter — APPROVED** [review]

**Comment by proux01** [comment]
> Hi @proux01 sorry for the delay, can you rebase?

@ejgallego done

File: vernac/record.ml [hunk 4]
```diff
@@ -703,15 +720,28 @@ let declare_structure { Record_decl.mie; primitive_proj; impls; globnames; globa
...
+
+(* declare definitional class (typeclasses that are not record) *)
+let declare_class_constant ~univs paramimpls params data =
+  let id, _, rdata, is_coercion, coers, inhabitant_id = get_class_params data in
+  assert (not is_coercion);  (* should be ensured by caller *)
+  let implfs = rdata.DataR.implfs in
+  let field, binder, proj_name = match rdata.DataR.fields with
+    | [ LocalAssum ({binder_name=Name proj_name} as binder, field)
+      | LocalDef ({binder_name=Name proj_name} as binder, _, field) ] ->
+       let binder = {binder with binder_name=Name inhabitant_id} in
```
Comment (ejgallego) [inline vernac/record.ml:742]: Same identation problem.

File: vernac/record.ml [hunk 5]
```diff
@@ -582,6 +583,17 @@ let extract_record_data records =
   in
   ps, data
 
+(* Returns [None] for regular records, [Some false] for regular typeclasses
+   (records) and [Some true] for definitional typeclasses. *)
```
Comment (ejgallego) [inline vernac/record.ml:587]: Maybe introduce a type to make this clearer.

File: vernac/record.ml [hunk 6]
```diff
@@ -597,7 +609,37 @@ let pre_process_structure udecl kind ~template ~cumulative ~poly ~primitive_proj
         typecheck_params_and_fields (kind = Class true) poly udecl ps data) ()
   in
   let template = template, auto_template in
-  template, impargs, params, univs, variances, data
+  let adjust_impls impls = match kind_class kind with
+    | None -> impargs @ [CAst.make None] @ impls
+    | Some _ -> implicits_of_context params @ impls in
+  let data = List.map (fun ({ DataR.implfs; _ } as d) -> { d with DataR.implfs = List.map adjust_impls implfs }) data in
+  (* let map (min_univ, arity, fieldimpls, fields) { Ast.name; is_coercion; cfs; idbuild; _ } = *)
```
Comment (ejgallego) [inline vernac/record.ml:616]: Left-over comment.

File: vernac/record.ml [hunk 7]
```diff
@@ -582,6 +583,17 @@ let extract_record_data records =
...
+let kind_class =
+  let open Vernacexpr in
+  function Class def -> Some def
+  | Inductive_kw | CoInductive | Variant | Record | Structure -> None
+
+let implicits_of_context ctx =
+  List.map (fun name -> CAst.make (Some (name,true)))
+    (List.rev (Anonymous :: (List.map RelDecl.get_name ctx)))
+
 let pre_process_structure udecl kind ~template ~cumulative ~poly ~primitive_proj finite (records : Ast.t list) =
```
Comment (ejgallego) [inline vernac/record.ml:597]: Let's take this occassion to document this function.

File: vernac/record.ml [hunk 8]
```diff
@@ -703,15 +720,28 @@ let declare_structure { Record_decl.mie; primitive_proj; impls; globnames; globa
...
-
-let build_class_constant ~univs ~rdata ~primitive_proj field implfs params paramimpls coers binder id proj_name =
+let get_class_params = function
+  | [{ Data.id; idbuild; rdata; is_coercion; coers; inhabitant_id }] ->
+     let coers = List.map (fun { Data.pf_subclass; pf_priority } -> if pf_subclass then Some { hint_priority = pf_priority; hint_pattern = None } else None) coers in
+     id, idbuild, rdata, is_coercion, coers, inhabitant_id
+  | _ ->
+     CErrors.user_err (str "Mutual definitional classes are not supported.")
+
+(* declare definitional class (typeclasses that are not record) *)
```
Comment (ejgallego) [inline vernac/record.ml:734]: Please let's profit and document the pre-conditions here.

**Review by ejgallego — APPROVED** [review]
Seem like an improvement to me, but the code is still in quite terrible shape. I wonder if instead of making more code common, we could separate the paths more, but have a more granular component view and have the separate paths share code relative to these common components.

I have left a few comments if you wanna do another pass.

File: vernac/record.ml (hunk 7 above)
Comment (proux01) [reply vernac/record.ml:597]: Unfortunately, I don't master the details of this function (only moved it around). I'd rather not write anything rather than a misleading comment.

File: vernac/record.ml [hunk 9]
```diff
@@ -597,7 +609,37 @@ let pre_process_structure udecl kind ~template ~cumulative ~poly ~primitive_proj
         typecheck_params_and_fields (kind = Class true) poly udecl ps data) ()
   in
   let template = template, auto_template in
-  template, impargs, params, univs, variances, data
+  let adjust_impls impls = match kind_class kind with
+    | None -> impargs @ [CAst.make None] @ impls
+    | Some _ -> implicits_of_context params @ impls in
+  let data = List.map (fun ({ DataR.implfs; _ } as d) -> { d with DataR.implfs = List.map adjust_impls implfs }) data in
+  (* let map (min_univ, arity, fieldimpls, fields) { Ast.name; is_coercion; cfs; idbuild; _ } = *)
+  let map rdata { Ast.name; is_coercion; cfs; idbuild; default_inhabitant_id; _ } =
```
[the remark this answers is not in the document]
Comment (proux01) [reply vernac/record.ml:616]: Well it's used in a `List.map2` so I'd find it a bit strange to have a separate function requiring two list of the same length.

File: vernac/record.ml (hunk 8 above)
Comment (proux01) [reply vernac/record.ml:734]: Done

**Comment by proux01** [comment]
Thanks @ejgallego for the careful review. Your comments should now be taken into account.
I agree it's still far from perfect. Note that situation should improve a bit when the whole #15802, #16224, #16230 thing will be sorted out in a few releases.

**Comment by ejgallego** [comment]
Great! I will merge as soon as the CI pass.

**Comment by ejgallego** [comment]
@coqbot: merge now
